### ISSUE TITLE
Warn about bash history expansion issue

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#1849]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
git clean -Xdn . --exclude="!.idea !preserve"

# will run: bash: !.idea: event not found
```

#### Here's what shellcheck currently says:

```console
$ shellcheck myscript
No issues detected!

$
```

#### Here's what I wanted or expected to see:

```sh
git clean -Xdn . --exclude="!.idea !preserve"
                            ^ [SC9999]: Single quote or escape this, unless you are trying to receive the previous command containing '.idea'.
```

--------

https://www.gnu.org/software/bash/manual/html_node/History-Interaction.html

> History expansion takes place in two parts. The first is to determine which line from the history list should be used during substitution. The second is to select portions of that line for inclusion into the current one. The line selected from the history is called the event, and the portions of that line that are acted upon are called words. Various modifiers are available to manipulate the selected words. The line is broken into words in the same fashion that Bash does, so that several words surrounded by quotes are considered one word. History expansions are introduced by the appearance of the history expansion character, which is ‘!’ by default.
> 
> History expansion implements shell-like quoting conventions: a backslash can be used to remove the special handling for the next character; single quotes enclose verbatim sequences of characters, and can be used to inhibit history expansion; and characters enclosed within double quotes may be subject to history expansion, since backslash can escape the history expansion character, but single quotes may not, since they are not treated specially within double quotes.
> 
> When using the shell, only ‘\’ and ‘'’ may be used to escape the history expansion character, but the history expansion character is also treated as quoted if it immediately precedes the closing double quote in a double-quoted string.

**Comment by gromgit** [comment]
Have you tried running the snippet you posted _as a script_?

History expansion is **disabled** by default in scripts. You'd only get that error in interactive mode (read: command line) and/or with a deliberate `set -H` (highly unusual in a script).

**Comment by stdedos** [comment]
I didn't; I have run it as part of `make clean` recipe though.

I don't know what's the difference between `make` calling commands and interactive mode, but I'd expect it'd be closer to script execution.

**Comment by gromgit** [comment]
What does your Makefile's `clean` target look like? Because I can't replicate your problem:
```
$ cat Makefile
test:
        git clean -Xdn . --exclude="!.idea !preserve"

$ make
git clean -Xdn . --exclude="!.idea !preserve"
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
make: *** [Makefile:2: all] Error 128
```
In any event, history expansion as you described it simply _can't_ happen in scripts without deliberate and conscious action on the developer's part.

**Comment by gromgit** [comment]
> I don't know what's the difference between `make` calling commands and interactive mode, but I'd expect it'd be closer to script execution.

For GNU make, it's basically:
```
${SHELL} ${.SHELLFLAGS} ${RECIPE}
```
where `SHELL` and `.SHELLFLAGS` are make's own variables. So a typical execution would be:
```
# SHELL="/bin/bash" .SHELLFLAGS="-c"
/bin/bash -c 'git clean -Xdn . --exclude="!.idea !preserve"'
```
which the bash man page categorically classifies as non-interactive (read: no history expansion).

**Comment by stdedos** [comment]
I guess you are right:

```bash
echo "'/bin/bash' '-c' ''"
'/bin/bash' '-c' ''
git clean -Xdff . --exclude="!.idea !preserve"
```
